enterprise/frontend/src/metabase-enterprise/sandboxes/components/LoginAttributesWidget/LoginAttributesWidget.unit.spec.tsx

```diff
@@ -0,0 +1,118 @@
+import { userEvent } from "@testing-library/user-event";
+
+import { setupUserEndpoints } from "__support__/server-mocks";
+import { renderWithProviders, screen } from "__support__/ui";
+import { Form, FormProvider } from "metabase/forms";
+import type {
+  StructuredUserAttributes,
+  UserAttributeMap,
+} from "metabase-types/api";
+import { createMockUserListResult } from "metabase-types/api/mocks";
+
+import { LoginAttributesWidget } from "./LoginAttributesWidget";
+
+const changeInput = async (fromValue: string, toValue: string) => {
+  const input = await screen.findByDisplayValue(fromValue);
+  await userEvent.clear(input);
+  await userEvent.type(input, toValue);
+};
+
+const simpleAttributes: UserAttributeMap = {
+  "@tenant.slug": "bug_gym",
+  color: "green",
+  type: "insect",
+  personal: "secret",
+};
+
+const structuredAttributes: StructuredUserAttributes = {
+  type: {
+    // overridden tenant attribute
+    source: "user",
+    frozen: false,
+    value: "insect",
+    original: {
+      source: "tenant",
+      frozen: false,
+      value: "bug",
+    },
+  },
+  color: {
+    // inherited tenant attribute
+    source: "tenant",
+    frozen: false,
+    value: "green",
+  },
+  personal: {
+    // personal attribute
+    source: "user",
+    frozen: false,
+    value: "secret",
+  },
+  "@tenant.slug": {
+    // immutable tenant slug
+    source: "system",
+    frozen: true,
+    value: "bug_gym",
+  },
+};
+
+const setup = () => {
+  const onChange = jest.fn();
+  const onSubmit = jest.fn();
+  const onError = jest.fn();
+
+  setupUserEndpoints(
+    createMockUserListResult({
+      structured_attributes: structuredAttributes,
+    }),
+  );
+
+  renderWithProviders(
+    <FormProvider
+      initialValues={{ login_attributes: simpleAttributes }}
+      onSubmit={onSubmit}
+    >
+      {({ values }) => (
+        <Form>
+          <LoginAttributesWidget name="login_attributes" userId={1} />
+          <pre aria-label="values-output">{JSON.stringify(values)}</pre>
```

Comment:
this is a little silly, but I couldn't think of a better way to access the form values